### PR TITLE
Scheduler rewrite, improvements, debugging, and tests

### PR DESCRIPTION
**Pull request by QuinnDamerell**
I wanted to get involved with the project so I pulled the code and found the scheduler as a good starting point. As I looked at the code I realized it was quite over complicated and there seemed to be two different concepts going on in the code. There is a static array of tasks that can be ran that is compiled into the code and there was also a queue of tasks, but the tasks were only added to the queue on boot and never changed.

I simplified the logic to only use the task list, which makes the code much easier to maintain but remove the possibility of queuing random task at any given time. That can still be added later down the road, but it didn’t seem like it was really a feature that would be needed in the system.
I also made some improvements, the scheduler had the concept of a realtime guard (tasks would not be scheduled if they would run past the time when a realtime task needed to be scheduled) but the logic had a bug that would only make it apply if the realtime events were first in the queue. It also only considered the highest running time of all of the tasks, so if there was one task that was super long it wouldn’t let any tasks queue in the time slot. I changed the logic to work for all realtime tasks, and to also consider each task length individually so if one will fit it will run.

I also improved the timing period for time based tasks, so if a task should run every 100ms the period between runs should be much closer to 100ms. The old code would reschedule the task to run 100ms from when it was executed, which means if the task was executed at 112 it would be queued to run at 212 next. A better way of doing this is to always schedule to the time it should have ran, so if it executed at 112 it would be scheduled to run at 200 next. The result is the time tasks are much more consistent. 

And finally, I added a ton of unit tests. There were unit tests for the queue in the scheduler, but none for any of the scheduling logic. I added unit test for just about everything, including all of the realtime guard and priority features.

### CODE REVIEW

**Comment by hydra** [comment]
ahh very interesting! I will have a look at this soon for sure!

**Comment by hydra** [comment]
@digitalentity @martinbudden @ledvinap would love your comments on this if you have time.

**Comment by QuinnDamerell** [comment]
Great! I would love feedback on it, feel free to give me any thoughts you have. You can get a good feel of whats going on / what I thought the important code paths are by looking at the unit tests.

Random question, is there a reason why the code hasn't been migrated toward c++ STL? I'm so accustomed to it now I feel old school writing more classic c. Lol.

**Comment by hydra** [comment]
i have heard from many developers that they don't want C++.  Personally I love the encapsulation you get with C++ and I feel it would make developers think harder about the code they write rather then just hacking a new file and collection of methods and global variables up like you get with C.  It's a discussion for another issue though, let's keep this discussion on this PR focussed.

**Comment by QuinnDamerell** [comment]
Yeah I agree with you 100%, but it is an issue for another time.

**Comment by proggod** [comment]
C is much more efficient then c++, and quite honestly c++ is dying and being replaced by better languages.   When your writing bare metal code you want to be as close to possible to native machine language. Since no one wants to write in assembly c is the next best choice.

**Comment by proggod** [comment]
I don't like the dead wait loop

**Comment by QuinnDamerell** [comment]
@proggod , which dead loop? For the main execute function?

**Comment by QuinnDamerell** [comment]
(pushed to pull in the unit test fix)

**Comment by proggod** [comment]
There was a while, inside the while loop.   It shouldn't be a big deal, but while loops without a potential to exit have caused problems in the code before.   I will look at your code more in detail tomorrow, but it looked pretty good to me.   You should join slack, I think cleanflight could use some more help if you have time.

**Comment by proggod** [comment]
Also you should try and fix your errors, so a clean merge could be done.

**Comment by QuinnDamerell** [comment]
Sure, whats the slack address? 

I pulled master and committed the merge that fixes the unit test fail, if there is another problem I will take a look tomorrow. 😁

**Comment by proggod** [comment]
http://www.raceflight.net/invite join there, and message me and I can give you some more details :)

Ya the errors didn't seem to be yours, I'll look into them.

**Comment by martinbudden** [comment]
[Off topic - for the record I also prefer C++, and for embedded code you can avoid the primary inefficiencies of C++ vs C]

And @QuinnDamerell , it's always good to have someone new get involved in the project.

Back to the scheduler changes:

> I simplified the logic to only use the task list, which makes the code much easier to maintain but remove the possibility of queuing random task at any given time.

The priority queue was introduced as an efficiency measure, not for it's ability to queue tasks at a given time. It allows three main efficiencies:

1. The realtime check loop only iterates through the realtime tasks, not all the tasks - possible since the tasks are in priority order.
2. processing tasks in priority order means that the conditional statements more often execute in our favour (that is take the quickest condition),
3. it removes the need for the `isEnabled` flag - if a task is enabled it is in the queue.  So only enabled tasks are iterated through

> but the logic had a bug that would only make it apply if the realtime events were first in the queue. 

Not a bug, a deliberate optimisation. The queue is a priority queue, so realtime events **are** the first in the queue.

> I also improved the timing period for time based tasks, so if a task should run every 100ms the period between runs should be much closer to 100ms. 

Jitter reduction is sometimes welcome, sometimes not. For example for sonar there must be at least 70ms between readings, otherwise the readings can be erroneous, so for sonar jitter reduction is not good. Also jitter reduction should not come at the cost of performance.

There is a comment in the code:
```
// We shouldn't select this task because it will take too long and the realtime task won't be ran on time.
// However we don't want to 100 starve the task, so we will heavily weight the task, but if it still wins the
// starvationPriority game it will be ran anyways. The weight will be higher for lower priority tasks.
```
A starved task should **never** pre-empt the realtime task.

The way an overloaded CPU is handled is that we prevent the FC from arming if the CPU is overloaded. That way the user can disable some features and try again. In practice this is only an issue with the F1 board (ie primarily NAZE and CC3D)

Note that this is not a general purpose scheduler, it is a scheduler specific to the needs of a Cleanflight.

**Comment by digitalentity** [comment]
@QuinnDamerell 
Primary task for this scheduler is being fast, being maintainable is the secondary objective. Priority queue approach might seem complex, but it improves performance. Your change is a step back - we've been using only one task list when I initially implemented the scheduler.

The only feisible change I see is jitter reduction, however as @martinbudden pointed out jitter reduction is not always desired and shouldn't come at expense of performance.

**Comment by borisbstyle** [comment]
I agree with @martinbudden. The scheduler is really mainly priority based.
Pretty much evolved superloop. If the schedulers goal was something else
RTOS would have been used.
The time interval between the tasks is sometimes crucial.
Also ESC writes are not allowed to occur too early as you can get
overlapping motor commands since it is just a pulse. Thats the way how the
current pid loop works.

I agree with @martinbudden. The scheduler is really mainly priority based
focused on realtime task. Pretty much evolved superloop. If the schedulers
goal was something else RTOS would have been used.
The time interval between the tasks is sometimes crucial.
Also ESC writes are not allowed to occur too early as you can get
overlapping motor commands since it is just a pulse.

More efficiency can be gained by minimising the busy wait on realtime task
for example and remove the delays. The penalty is just slightly getting out
of sync with gyro, which is totally not a big deal with higher cycletimes
now we usually run today when you do fast sampling where it doesnt matter.
Just have to make sure that you dont read the same sample twice, what can
be solved by using sampling fraction slower than gyro speed itself.

File: src/main/scheduler/scheduler.c [hunk 1]
```diff
@@ -32,247 +33,301 @@
...
-        if (taskQueueArray[ii] == task) {
-            memmove(&taskQueueArray[ii], &taskQueueArray[ii+1], sizeof(task) * (taskQueueSize - ii));
-            --taskQueueSize;
-            return true;
-        }
+#ifdef SCHEDULER_DEBUG_PRINT
+    printf("System Task Ran - avg system load (%d)\n",averageSystemLoadPercent);
+    if(isCpuOverloaded)
+    {
+        printf(" - SYSTEM IS OVERLAODED, WE SKIPPED TASK PERIODS.\n");
```
Comment (hydra) [inline src/main/scheduler/scheduler.c]: typo in OVERLOADED.

File: src/main/scheduler/scheduler.c [hunk 2]
```diff
@@ -32,247 +33,301 @@
...
-        }
+        task->isEnabled = (enabled && task->taskFunc);
     }
 }
 
-uint32_t getTaskDeltaTime(const int taskId)
+uint32_t getTaskTimeSinceLastRun(const int taskId)
 {
-    if (taskId == TASK_SELF || taskId < (int)taskCount) {
+    if ((taskId == TASK_SELF && currentTask != NULL) || (taskId < (int)taskCount && taskId >= 0)) {
```
Comment (hydra) [inline src/main/scheduler/scheduler.c]: i see this same 'if (..)' statement 3 times, perhaps it has a name and should be reused?

**Comment by ledvinap** [comment]
(repeating myself)
Here is tested, reasonably fast scheduler: https://github.com/ChibiOS/ChibiOS/blob/master/os/rt/src/chschd.c (and other in freeRTOS). 
When scheduling promises are updated (no priority groups, no scheduler based handling of starved tasks), it should be very easy to use in CF. And as a bonus migration to RTOS would be much easier in future...

File: src/main/scheduler/scheduler.h [hunk 3]
```diff
@@ -18,68 +18,99 @@
...
-    uint32_t     desiredPeriod;
-    uint8_t      staticPriority;
-    uint32_t     maxExecutionTime;
-    uint32_t     totalExecutionTime;
-    uint32_t     averageExecutionTime;
-    uint32_t     latestDeltaTime;
+    bool         isEnabled;                 // Indicates if the task is enabled.
+    uint32_t     desiredPeriod;             // the desired amount of time between each execution (in microseconds)
+    uint8_t      priority;                  // the priority of the task
+    bool         isWaitingToBeRan;          // Indicates if the task is waiting to be ran or not.
```
Comment (hydra) [inline src/main/scheduler/scheduler.h]: 'ran' is past tense, so 'isWaitingToBeRun'  but perhaps 'pending' might be simpler?

File: src/main/scheduler/scheduler.c [hunk 4]
```diff
@@ -32,247 +33,301 @@
...
+            } else {
+                const uint32_t newTimeInterval = nextExecuteAt - currentTime;
+                timeToNextRealtimeTask = MIN(timeToNextRealtimeTask, newTimeInterval);
+            }
         }
     }
-    const bool outsideRealtimeGuardInterval = (timeToNextRealtimeTask > realtimeGuardInterval);
+
+#ifdef SCHEDULER_DEBUG_PRINT
+    printf("Time Until Next Realtime (%u)\n", timeToNextRealtimeTask);
```
Comment (hydra) [inline src/main/scheduler/scheduler.c]: perhaps we should replace SCHEDULER_DEBUG_PRINT with DEBUG_PRINT(SCHEDULER, ...) macros or something?  not critical right now but I can see it being useful.

File: src/main/scheduler/scheduler.c [hunk 5]
```diff
@@ -32,247 +33,301 @@
...
-            if (task->taskAgeCycles > 0) {
-                task->dynamicPriority = 1 + task->staticPriority * task->taskAgeCycles;
-                waitingTasks++;
+            else
+            {
+                // This isn't event driven, see if it should be ran based on time. 
+                // Note it is important to use the time it should have been scheduled not when it was,
+                // this will give use more accurate interval times.
+                // TODO: handle currentTime rolling over.
+                if((task->lastIdealExecutionTime + task->desiredPeriod) <= currentTime)
```
Comment (hydra) [inline src/main/scheduler/scheduler.c]: can you add a unit test for this TODO item?

**Comment by hydra** [comment]
Can you rebase on master and fix the build problems?

Overall I find the code in the PR easy to follow.  There a few minor whitespace/style issues, such as the missing whitespace between `if` and the open brace `(` - use `if (` not `if(`.  Naming wise things seem pretty good.  There was a bit of logic duplication that I noted.  I will try out the unit tests and the code soon, over-all I'm liking the cleanup and I'm very in favour of cutting out stuff we don't need, fixing bugs and improving scheduling consistency so this is high priority for me.

**Comment by hydra** [comment]
some tasks would benefit in jitter reduction, yes.  perhaps it's possible to add a 'minimumInterval' property to such tasks and not run them earlier than that interval and elsewhere reduce jitter where possible?

**Comment by martinbudden** [comment]
When solving problems things should be made as simple as possible, no simpler.

Or in other words, problems have essential complexity and accidental complexity. The essential complexity is to do with the nature of the problem, the accidental complexity is to do with the nature of the solution and, as far as possible, should be eliminated.

Let's do some analysis. The problem is to schedule tasks with different priorities.

1. The number of tasks can vary at compile time (because of build flags like `BARO` etc)
2. The number of tasks can also at initialisation time (because of the user switching features on and off in the configurator).
3. After initialisation the number of tasks cannot change.

(1) means we need an iterable container for the tasks. An array is the simplest type of container we can use.
(2) means that we need an additional form of selection. There are two obvious solutions - an enable flag on each task, or a queue of enabled tasks
(3) means that if we choose a queue of enabled tasks it can be implemented as a simple buffer, if the number of tasks could vary arbitrarily at runtime then a linked list should be used for the queue.

Task-scheduling is a well understood problem and nothing above is new.

Using a flag on each task is sub-optimal, for a number of reasons that I have already stated in this thread.

Using a priority queue for task scheduling is pretty well the canonical implementation and as such is well understood and therefore easy to maintain.

There is room for improvement in the  current solution (as there is room for improvement in all code) but it will outperform the proposed changes.

I believe the proposed change is an oversimplification of the problem. That is essential aspects of the problem have been overlooked in the goal of simplification. It will result is less performant code.

@hydra I'm not sure why you want to prioritise replacing a working scheduler just before a release. If you don't like the existing code, fair enough, but surely it is wise to wait until after the current release to replace it.

Also, if you don't like the priority queue, I don't understand why you merged in PR https://github.com/cleanflight/cleanflight/pull/1784 - Scheduler optimized by using priority queue. This PR basically reverts those changes.

**Comment by ledvinap** [comment]
@martinbudden : Another possibility is to use linked list of tasks.
Each module can allocate its own task description structure and register (multiple) tasks in initialization routine. No need for global structure separated from modules (it is possible to use linker sections to get task descriptors in continuous memory if necessary)

Run queue and wait queue may be easily separated, first sorted by priority, second by time to run.
Task insertion will be O(n) -  using O(log n) is probably overkill, there is quite few tasks so simpler algorithm will be faster.
This will probably improve performance compared to current behavior - no need to scan tasks on each scheduler call (or maybe performance will be similar - one task insert per schedule instead of one scan).

It will be easy to schedule task from event - no need for slow checker functions

It will allow much easier transition to RTOS in future if desired

**Comment by martinbudden** [comment]
@ledvinap , yes a linked list of tasks is another possibility. And RTOS is a whole other discussion...

But I'm generally not in favour of making changes in this area, especially just before a release:
1. The existing code works and is performant
2. Any bugs in the scheduler code are pretty sure to impact the user
3. Any work on the scheduler is an opportunity cost, when there are quite a few more important and urgent things to do. Migration to the new IO, for example.

**Comment by hydra** [comment]
@martinbudden I have largely been leaving the scheduler side of things to those with an interest in it.  After these comments it's clear to me that the proposed changes are probably not right for now.

Also now agreed that due to the comments above it's not right to merge or change things for 1.14.0.

Agreed that each module should be able to add/register one or more tasks. this will help with modularisation for sure and will likely remove some FC specific code from the tests which is a good thing.

Agreed that we should do the new IO first.

OK, let's park this until after 1.14.0.

@QuinnDamerell I hope you are not disheartened by any of this, i think the goals are clearer now.  Definately include @martinbudden @digitalentity @borisbstyle and @ledvinap before embarking on writing new scheduler changes.  Perhaps you can think of a plan based on all the above?

Also, if there's any bugs in the current scheduler code (as you noted) perhaps we should fix them now for the 1.14.0 release?

**Comment by QuinnDamerell-MS** [comment]
@hydra yeah sure!

Sorry I had things pick up at work so I haven't had time to read through all of the discussion. But when I get sometime I will look through it and respond. :)

I'm fine with pushing the work off / refactoring the changes. I will sync with the rest of the contributors and see what changes we want to have made and the best ways to do them. :)

**Comment by martinbudden** [comment]
@QuinnDamerell , and yes, despite my criticism, I hope you are not discouraged. My criticism is intended to be constructive and is of the code, not the person writing it. The scheduler is at the heart of the code, so any bugs introduced will manifest themselves, and they can do so in insidious and hard to track down ways.

**Comment by hydra** [comment]
@martinbudden indeed.

@QuinnDamerell great, looking forward to seeing what you come up with.

**Comment by hydra** [comment]
@QuinnDamerell any progress?

**Comment by hydra** [comment]
closing, lack of progress and developer response.

**Pull request closed without merging**